### DISCUSSION TITLE
Should I enable Windows Insider Preview Builds on UWP?

### DISCUSSION THREAD

**Discussion by gganjang** [microsoft/msquic#1428] [Q&A] [answered]
I'm working on quic client targeting MS Hololens 2 which uses UWP for OS.

My questions are:
1. Should I enable Windows Insider Preview Builds on Hololens2 for Schannel?
--> if so, is it possible to use TLS 1.3 on UWP? (this question is probably for hololens team..)

2. Should I turn on UWP-related option in cmakeLists when building MsQUIC library with x86_arm??
--> if so, what's the effects of UWP option??

3. What's the reason why UWP option does not come along with OpenSSL?? (probably related to the question No.1)

Anyway, I'm pretty sure that MsQuic's going to work well with Hololens 2, but just wanted to be confirmed as much as possible...

**Comment by ThadHouse** [comment, marked as answer]
I'll try and get an answer to 1.

For 2, UWP is missing a few APIs and links to a different set of libraries. The UWP setting just make the code work with that configuration.

For 3, OpenSSL configured for UWP would work, we just haven't had an ask for it yet, so I was not able to try it. In theory you should just be able to remove the check that disables it, and change `VC-WIN32` to `VC-WIN32-ONECORE` and `VC-WIN64A` to `VC-WIN64A-ONECORE`. Then the build should work. It might also require removing `no-shared` from the OpenSSL build arguments, I don't remember if static builds work correctly. But in theory nothing fundamental is broken with UWP and OpenSSL, the build would just have to be changed slightly to make it work.

**Reply by nibanks** [reply]
For 1, I'd recommend just trying it out and seeing what happens. All the UWP platforms have the same core Windows OS code, but the various insider releases might be on different cadences, and it's hard for us to keep track of them.

**Comment by ThadHouse** [comment]
To add on to this, full support for UWP builds for all platforms was added a few weeks ago. Additionally, all restricted API's were worked around (for OpenSSL builds at least) so UWP apps should be publishable to the store, in addition to working on all UWP based platforms. I don't have a hololens, but it should work.

**Comment by gganjang** [comment]
Update!
MsQuic works on Hololens 2.

Since I have to finish my wrapper library first, I've not been able to test all functions yet. However, I confirmed that native UWP runs well with 'MsQuicOpen' API at least. I built MsQuic targeting 'arm' architecture with UWP build option 'ON' in CMakeLists. The native UWP app linked msquic.lib only.

Using .dll for UWP app is little tricky which made me spend over a week. Under binary directory after build is successful, there is a directory named 'AppX'. If you open the directory you can find another .exe file in it even if there is an .exe outside of 'AppX' directory. Put msquic.dll in 'AppX', then everything works perfect.
Reference: https://forum.unity.com/threads/trouble-loading-uwp-native-c-dll.524011/
